File: rximagepicker/src/main/java/com/qingmei2/rximagepicker/ui/ActivityPickerViewController.java
package com.qingmei2.rximagepicker.ui;

import android.app.Activity;
import android.content.Intent;
import android.support.v4.app.FragmentActivity;

import com.qingmei2.rximagepicker.entity.Result;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.subjects.PublishSubject;

public final class ActivityPickerViewController implements ICustomPickerView {

    private volatile static ActivityPickerViewController INSTANCE;

    private PublishSubject<Result> publishSubject;

    private Class<? extends Activity> activityClass;

    public static ActivityPickerViewController getInstance() {
        if (INSTANCE == null) {
            synchronized (ActivityPickerViewController.class) {
                if (INSTANCE == null) {
                    INSTANCE = new ActivityPickerViewController();
                }
            }
        }
        return INSTANCE;
    }

    private ActivityPickerViewController() {

    }

    public void setActivityClass(Class<? extends Activity> clazz) {
        activityClass = clazz;
    }

    public void resetSubject() {
        publishSubject = PublishSubject.create();
    }

    @Override
    public void display(FragmentActivity fragmentActivity,
                        int viewContainer,
                        String viewKey,
                        ICustomPickerConfiguration configuration) {
        resetSubject();
        fragmentActivity.startActivity(new Intent(fragmentActivity, activityClass));
    }

    @Override
    public Observable<Result> pickImage() {
        return publishSubject;
    }

    public void emitResult(Result result) {
        publishSubject.onNext(result);
    }

    public void emitResults(List<Result> results) {
        for (Result result : results) {
            emitResult(result);
        }
    }

    public void emitError(Throwable e) {
        publishSubject.onError(e);
    }

    public void endResultEmitAndReset() {
        publishSubject.onComplete();
        resetSubject();
    }
}

File: rximagepicker/src/main/java/com/qingmei2/rximagepicker/ui/ActivityPickerViewController.kt
package com.qingmei2.rximagepicker.ui

import android.app.Activity
import android.content.Intent
import androidx.fragment.app.FragmentActivity

import com.qingmei2.rximagepicker.entity.Result

import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject

class ActivityPickerViewController private constructor() : ICustomPickerView {

    private var publishSubject: PublishSubject<Result> = PublishSubject.create()

    private lateinit var activityClass: Class<out Activity>

    fun setActivityClass(clazz: Class<out Activity>) {
        activityClass = clazz
    }

    fun resetSubject() {
        publishSubject = PublishSubject.create()
    }

    override fun display(fragmentActivity: androidx.fragment.app.FragmentActivity,
                         viewContainer: Int,
                         configuration: ICustomPickerConfiguration?) {
        resetSubject()
        fragmentActivity.startActivity(Intent(fragmentActivity, activityClass))
    }

    override fun pickImage(): Observable<Result> {
        return publishSubject
    }

    fun emitResult(result: Result) {
        publishSubject.onNext(result)
    }

    fun emitResults(results: List<Result>) {
        for (result in results) {
            emitResult(result)
        }
    }

    fun emitError(e: Throwable) {
        publishSubject.onError(e)
    }

    fun endResultEmitAndReset() {
        publishSubject.onComplete()
        resetSubject()
    }

    companion object {

        @Volatile
        private var INSTANCE: ActivityPickerViewController? = null

        val instance: ActivityPickerViewController
            get() {
                if (INSTANCE == null) {
                    synchronized(ActivityPickerViewController::class.java) {
                        if (INSTANCE == null) {
                            INSTANCE = ActivityPickerViewController()
                        }
                    }
                }
                return INSTANCE!!
            }
    }
}
